core/unit_test/TestViewCtorProp.hpp

```diff
@@ -0,0 +1,145 @@
+
+  static_assert(std::is_same_v<decltype(Kokkos::view_alloc()), vcp_empty_t>);
+}
+
+// Check traits of base class Kokkos::Impl::ViewCtorProp<void, std::string>.
+TEST(TEST_CATEGORY, vcp_label_base_traits) {
+  using vcp_label_base_t = Kokkos::Impl::ViewCtorProp<void, std::string>;
+
+  static_assert(std::is_same_v<typename vcp_label_base_t::type, std::string>);
+
+  // Check that the base class is default constructible. The default constructor
+  // may be called by the copy constructor of deriveded classes, such as when
+  // copy constructing a view constructor properties object from another view
+  // constructor properties object that holds fewer properties.
+  static_assert(std::is_default_constructible_v<vcp_label_base_t>);
+
+  static_assert(std::is_constructible_v<vcp_label_base_t, std::string>);
+  static_assert(std::is_constructible_v<vcp_label_base_t, const std::string&>);
+  static_assert(std::is_constructible_v<vcp_label_base_t, std::string&&>);
+
+  constexpr unsigned N = 3;
+  static_assert(std::is_constructible_v<vcp_label_base_t, const char[N]>);
+  static_assert(std::is_constructible_v<vcp_label_base_t, char[N]>);
+
+  static_assert(std::is_constructible_v<vcp_label_base_t, char*>);
+}
+
+// Check traits of derived class Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_traits) {
+  using vcp_label_base_t = Kokkos::Impl::ViewCtorProp<void, std::string>;
+  using vcp_label_t      = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  static_assert(std::is_base_of_v<vcp_label_base_t, vcp_label_t>);
+
+  static_assert(vcp_label_t::has_label);
+
+  // Check that the derived class is not default constructible. It is a design
+  // choice to not allow the default constructor to be called.
+  static_assert(!std::is_default_constructible_v<vcp_label_t>);
+
+  static_assert(std::is_constructible_v<vcp_label_t, std::string>);
+  static_assert(std::is_constructible_v<vcp_label_t, const std::string&>);
+  static_assert(std::is_constructible_v<vcp_label_t, std::string&&>);
+
+  constexpr unsigned N = 3;
+  static_assert(std::is_constructible_v<vcp_label_t, const char[N]>);
+  static_assert(std::is_constructible_v<vcp_label_t, char[N]>);
+
+  static_assert(std::is_constructible_v<vcp_label_t, char*>);
+}
+
+// Check that the constructor of Kokkos::Impl::ViewCtorProp<std::string>
+// moves a label passed by rvalue reference.
+TEST(TEST_CATEGORY, vcp_label_constructor_can_move) {
+  using vcp_label_t = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  std::string label("our label");
+
+  vcp_label_t prop(std::move(label));
+
+  ASSERT_TRUE(label.empty());
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+}
+
+// Check that Kokkos::view_alloc moves a label passed by rvalue reference.
+TEST(TEST_CATEGORY, vcp_label_view_alloc_can_move) {
+  std::string label("our label");
+
+  auto prop = Kokkos::view_alloc(std::move(label));
```

Comment:
This is equivalent to L100 in the previous unit test.  What am I missing?

Reply:
One test checks move semantics with the internals. The other one goes through the public API `Kokkos::view_alloc`.

I would argue that the second test could be enough. Indeed, if `auto prop = Kokkos::view_alloc(std::move(label));` does not move, it means that either `view_alloc` does not perfect forward, or that the first test should also fail.

Reply:
Indeed. We merged the two tests into a single tests. The combined tests now checks both `view_alloc` (public API) and the constructor (in the `Impl` namespace).